### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by Jeadie**
> This PR build off of #60, it will be easier to read once it's merged.

### Improvements
- Basic, extendable support to use any LLM model you like
- Better typing, structs, enums etc. to make the code more understandable.
- Better package layout

### CODE REVIEW

File: .gitignore
Comment (patillacode) [inline .gitignore:2]: Are you sure we don't want to leave `main_prompt`?
maybe we should use a `.keep` file or specify in the `gitignore` that we want to ignore everything EXCEPT for the `main_prompt` file

**Review by patillacode — CHANGES_REQUESTED** [review]
Hi @Jeadie 

Thanks for your contribution, as you can see I just merged #60 

In order to merge this PR with the improvements please:
- rebase from `main` (all extra commits should disappear)
- fix conflicts

Also, I left a comment!

File: .gitignore
Comment (jebarpg) [reply .gitignore:2]: Leaving the main_prompt out would help with reducing conflicts when people are testing or just using the repo. Every time you go to do a git pull to get the latest update you'll end up with merge conflicts... At least that's my 2¢

Comment (patillacode) [reply .gitignore:2]: Fair enough, I tend to agree.

Users will only use the example once to try, so yeah, no biggie.

**Comment by patillacode** [comment]
Thanks @Jeadie 

@AntonOsika I would like your confirmation on this one before I merge.

**Comment by AntonOsika** [comment]
Thanks for the good work keeping this up with main @Jeadie – and wanting to get it merged!

I added feedback on how we can get this in a really good update.

What I also propose:
- We first merge the changes to the types
- Then we look at the PR that introduce some complexity so that we can have multiple types of LLMs! (which will be very useful in the future)

Would you be able to look through my comments, see if you agree with the proposed dataclassses, and if so create a new PR that only updates the types?

**Comment by AntonOsika** [comment]
I realised we can just use `json.dumps(dataclasses.asdict())`, no need for additional dependencies 👌

**Review by AntonOsika — CHANGES_REQUESTED** [review]

**Comment by Jeadie** [comment]
I had the growing suspicion as I worked on this that `Tuple[Role, Message]` should become a complete dataclass. I started this change to make the function signatures more understandable and typed, and to be more robust, when/if a Role or Message get more complex.

File: gpt_engineer/models.py
Comment (Jeadie) [inline gpt_engineer/models.py:7]:
```suggestion
```

**Comment by Jeadie** [comment]
I think I will move the support for mulitple models to a new PR, just focus this one on improving typing etc.

File: gpt_engineer/steps.py
Comment (Jeadie) [reply gpt_engineer/steps.py:29]: Sorry, could you elaborate on this, I don't understand. 
```python
user_msg = dbs.input["main_prompt"]
messages = [(Role.SYSTEM, Message(dbs.identity["qa"]))]
```
Is needed before the while loop because updates are not equivalent.

File: gpt_engineer/steps.py
Comment (Jeadie) [reply gpt_engineer/steps.py:35]: Sorry, I didn't even realise i change much here.

Comment (Jeadie) [reply gpt_engineer/steps.py:35]: Oh i think this was an old difference.

File: gpt_engineer/steps.py
Comment (Jeadie) [reply gpt_engineer/steps.py:38]: Correct, I made an error in ai.next that masked this.

File: pyproject.toml
Comment (Jeadie) [reply pyproject.toml:22]: Subpackage, but as mentioned will remove other LLM models from this

**Comment by AntonOsika** [comment]
Ping me when I should review!

**Comment by AntonOsika** [comment]
Curious about progress @Jeadie

**Comment by AntonOsika** [comment]
My idea of introducing dataclass of Messages needs quite some work with appending messages so I'd be happy to reconsider

**Comment by AntonOsika** [comment]
Hey! Good ideas here. I'm not sure if we should merge or if we should use the ideas and do a new PR altogether. For example, maybe we want to put messages in its own file and give it classmethods like "to_dict" etc.

Before merging, would be good to look at how e.g. langchain and others handles this.

**Comment by Jeadie** [comment]
I agree

**Comment by Jeadie** [comment]
Looking into Langchain, this is their data structures for their messages, [source](https://github.com/hwchase17/langchain/blob/master/langchain/schema.py#L85).
```python
class BaseMessage(Serializable):
    """Message object."""

    content: str
    additional_kwargs: dict = Field(default_factory=dict)

    @property
    @abstractmethod
    def type(self) -> str:
        """Type of the message, used for serialization."""

    @property
    def lc_serializable(self) -> bool:
        """This class is LangChain serializable."""
        return True
```
Subclasses are for various use cases, for example: HumanMessage, AIMessage, FunctionMessage. These messages are based off a `Serializable` class that inherits from pydantic and an abstract class [source](https://github.com/hwchase17/langchain/blob/master/langchain/load/serializable.py#L33). 

It appears to be a similar approach. One difference is the use of a type property instead of a separate type/role enum. For example, in LangChain
```python
@property
def type(self) -> str:
    """Type of the message, used for serialization."""
    return "human"
```

**Comment by Jeadie** [comment]
These then get used internally per specific use cases, which provide public methods and uses these message types internally. For example ([source](https://github.com/hwchase17/langchain/blob/master/langchain/schema.py#L307C13-L352))
```python
class BaseChatMessageHistory(ABC):
    """Base interface for chat message history
    See `ChatMessageHistory` for default implementation.
    """

    """
    Example:
        .. code-block:: python

            class FileChatMessageHistory(BaseChatMessageHistory):
                storage_path:  str
                session_id: str

               @property
               def messages(self):
                   with open(os.path.join(storage_path, session_id), 'r:utf-8') as f:
                       messages = json.loads(f.read())
                    return messages_from_dict(messages)

               def add_message(self, message: BaseMessage) -> None:
                   messages = self.messages.append(_message_to_dict(message))
                   with open(os.path.join(storage_path, session_id), 'w') as f:
                       json.dump(f, messages)
               
               def clear(self):
                   with open(os.path.join(storage_path, session_id), 'w') as f:
                       f.write("[]")
    """

    messages: List[BaseMessage]

    def add_user_message(self, message: str) -> None:
        """Add a user message to the store"""
        self.add_message(HumanMessage(content=message))

     ...
```

**Comment by AntonOsika** [comment]
Makes sense.

We really value simplicity so I am inclined to u turn in the name of simplicity here and just have a Message class and use lists like you suggested initially.

**Comment by AntonOsika** [comment]
I.e., just merging this part first!^

Would be an improvement for sure
image

**Comment by AntonOsika** [comment]
Thanks for the contributions @Jeadie helped us realised what we want!

Moving discussion to #466

Will close since it's merge, feel free to reopen.

**Pull request closed without merging**
